domain/status/state/storage.go

```diff
@@ -0,0 +1,335 @@
+		StatusID:       statusID,
+		Message:        sts.Message,
+		UpdatedAt:      sts.Since,
+	}
+	stmt, err := st.Prepare(`
+INSERT INTO storage_filesystem_status (*) VALUES ($filesystemStatusInfo.*)
+ON CONFLICT(filesystem_uuid) DO UPDATE SET
+    status_id = excluded.status_id,
+    message = excluded.message,
+    updated_at = excluded.updated_at
+`, statusInfo)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, stmt, statusInfo).Run()
+	if err != nil {
+		return errors.Capture(err)
+	}
+	return nil
+}
+
+// SetVolumeStatus saves the given volume status, overwriting any
+// current status data. The following errors can be expected:
+// - [storageerrors.VolumeNotFound] if the volume doesn't exist.
+func (st *State) SetVolumeStatus(
+	ctx context.Context,
+	volumeUUID storage.VolumeUUID,
+	sts status.StatusInfo[status.StorageVolumeStatusType],
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Get current status.
+		currentStatus, isProvisioned, err := st.getVolumeProvisioningStatus(ctx, tx, volumeUUID)
+		if err != nil && !errors.Is(err, statusNotFound) {
+			return errors.Errorf("getting current volume status: %w", err)
+		}
+		if err == nil {
+			// Check we can transition from current status to the new status.
+			err = status.VolumeStatusTransitionValid(currentStatus, isProvisioned, sts)
+			if err != nil {
+				return errors.Capture(err)
+			}
+		}
+
+		return st.updateVolumeStatus(ctx, tx, volumeUUID, sts)
+	})
+	if err != nil {
+		return errors.Errorf("updating volume status for %q: %w", volumeUUID, err)
+	}
+	return nil
+}
+
+// getVolumeStatus gets the status of the given volume
+// and a bool indicating if it is provisioned.
+// The following errors can be expected:
+// - [storageerrors.VolumeNotFound] if the volume doesn't exist.
+func (st *State) getVolumeProvisioningStatus(
+	ctx context.Context,
+	tx *sqlair.TX,
+	uuid storage.VolumeUUID,
+) (status.StorageVolumeStatusType, bool, error) {
+	id := volumeUUID{
+		VolumeUUID: uuid.String(),
+	}
+	var sts storageProvisioningStatusInfo
+
+	stmt, err := st.Prepare(`
+SELECT    &storageProvisioningStatusInfo.*
+FROM      storage_volume sv
+LEFT JOIN storage_volume_status svs ON sv.uuid = svs.volume_uuid
+LEFT JOIN storage_instance_volume siv ON sv.uuid = siv.storage_volume_uuid
+WHERE     sv.uuid = $volumeUUID.uuid
+`, id, sts)
+	if err != nil {
+		return -1, false, errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, stmt, id).Get(&sts)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return -1, false, storageerrors.VolumeNotFound
+	} else if err != nil {
+		return -1, false, errors.Capture(err)
+	}
+	if !sts.StatusID.Valid {
+		return -1, false, statusNotFound
+	}
+	statusType, err := status.DecodeStorageVolumeStatus(int(sts.StatusID.Int16))
+	if err != nil {
+		return -1, false, errors.Capture(err)
+	}
+	isProvisioned := sts.StorageInstanceUUID != ""
```

Comment:
Just the same comments above re making this NULL able.

Reply:
yup